File: paper/make_docx.js
/**
 * Build the Word version of the paper in JavaScript.
 *
 * A LaTeX -> docx converter written directly against the OOXML model, rather
 * than shelling out to pandoc. It reads `main.tex` and the IEEE template and
 * emits `main-js.docx`.
 *
 * How it keeps the IEEE look: the template's `styles.xml` is lifted out of
 * `conference-template-a4.docx` and handed to the `docx` library as
 * `externalStyles`, so paragraphs can reference IEEE's own style ids
 * (`papertitle`, `Abstract`, `figurecaption`, ...) and inherit their exact
 * fonts and spacing. The page is set to A4 with a single-column front matter
 * section followed by a two-column body, which is how IEEE lays the format out.
 *
 * Known limitation, stated rather than hidden: maths is rendered as Unicode
 * text, not as native Word equations. The pandoc route (`make_docx.py`)
 * produces real OMML equation objects and is the better choice when the maths
 * matters. In this paper the maths is simple enough --- superscripts,
 * subscripts, Greek letters and inequalities --- that Unicode is faithful, but
 * a reader who wants to *edit* equations in Word should use the Python route.
 *
 * Usage:
 *   node paper/make_docx.js
 *   node paper/make_docx.js --single-column
 */

const fs = require("fs");
const path = require("path");
const JSZip = require("jszip");
const {
  Document, Packer, Paragraph, TextRun, ImageRun, Table, TableRow, TableCell,
  WidthType, AlignmentType, SectionType, BorderStyle,
} = require("docx");

const HERE = __dirname;
const TEX = path.join(HERE, "main.tex");
const TEMPLATE = path.join(HERE, "conference-template-a4.docx");
const FIGS = path.join(HERE, "figures");
const OUT = path.join(HERE, "main-js.docx");

/* ------------------------------------------------------------------ maths */
// LaTeX fragments -> Unicode. Applied to inline maths and to running text.
const SYMBOLS = [
  [/\\times/g, "×"], [/\\pm/g, "±"], [/\\leq/g, "≤"],
  [/\\geq/g, "≥"], [/\\approx/g, "≈"], [/\\ldots/g, "…"],
  [/\\dots/g, "…"], [/\\rightarrow/g, "→"], [/\\omega/g, "ω"],
  [/\\Omega/g, "Ω"], [/\\osc/g, "Ω"], [/\\evrate/g, "E_{tail}"],
  [/\\sim/g, "~"],
  [/\\%/g, "%"], [/\\&/g, "&"], [/\\_/g, "_"], [/\\#/g, "#"],
  [/---/g, "—"], [/--/g, "–"], [/``/g, "“"], [/''/g, "”"],
];
const SUP = { "0":"⁰","1":"¹","2":"²","3":"³","4":"⁴",
  "5":"⁵","6":"⁶","7":"⁷","8":"⁸","9":"⁹",
  "-":"⁻","−":"⁻","+":"⁺" };
const SUB = { "0":"₀","1":"₁","2":"₂","3":"₃","4":"₄",
  "5":"₅","6":"₆","7":"₇","8":"₈","9":"₉" };

function toUnicodeScript(s, table) {
  return [...s].every((c) => table[c]) ? [...s].map((c) => table[c]).join("") : null;
}

/** Render one inline-maths body (the text between `$`) as Unicode. */
function renderMath(m) {
  let s = m;
  for (const [re, to] of SYMBOLS) s = s.replace(re, to);
  s = s.replace(/\\mathrm\{([^}]*)\}/g, "$1").replace(/\\emph\{([^}]*)\}/g, "$1");
  s = s.replace(/\\text\{([^}]*)\}/g, "$1");
  // E_{\mathrm{tail}} -> E_tail ; x^{2} -> superscript where representable
  s = s.replace(/\^\{([^}]*)\}|\^(\S)/g, (_, a, b) => {
    const t = a !== undefined ? a : b;
    return toUnicodeScript(t, SUP) || "^" + t;
  });
  s = s.replace(/_\{([^}]*)\}|_(\S)/g, (_, a, b) => {
    const t = a !== undefined ? a : b;
    return toUnicodeScript(t, SUB) || "_" + t;
  });
  return s.replace(/[{}]/g, "").replace(/\\,/g, " ").trim();
}

/* ------------------------------------------------------- inline formatting */
/**
 * Turn a LaTeX paragraph into styled runs.
 * Handles \textbf, \emph/\textit, \texttt, \cite and inline maths; everything
 * else is flattened to plain text rather than dropped silently.
 */
// label -> the number its float actually carries, filled in before conversion.
const LABELS = {};

function runs(text, refs) {
  const out = [];
  // Order matters: maths first, so a `$...$` containing braces is not eaten
  // by the command matcher below.
  const re = /\$([^$]*)\$|\\(textbf|emph|textit|texttt)\{([^{}]*(?:\{[^{}]*\}[^{}]*)*)\}|\\cite\{([^}]*)\}|\\ref\{([^}]*)\}/g;
  let last = 0, m;
  const plain = (s) => {
    let t = s;
    for (const [r, to] of SYMBOLS) t = t.replace(r, to);
    return t.replace(/\\[a-zA-Z]+\s?/g, "").replace(/[{}]/g, "").replace(/~/g, " ");
  };
  while ((m = re.exec(text)) !== null) {
    if (m.index > last) out.push(new TextRun(plain(text.slice(last, m.index))));
    if (m[1] !== undefined) {
      out.push(new TextRun(renderMath(m[1])));
    } else if (m[2]) {
      const inner = plain(m[3]);
      if (m[2] === "textbf") out.push(new TextRun({ text: inner, bold: true }));
      else if (m[2] === "texttt") out.push(new TextRun({ text: inner, font: "Consolas" }));
      else out.push(new TextRun({ text: inner, italics: true }));
    } else if (m[4] !== undefined) {
      // Numeric citations, in the order the bibliography defines them.
      const nums = m[4].split(",").map((k) => refs.indexOf(k.trim()) + 1)
        .filter((n) => n > 0);
      out.push(new TextRun(`[${nums.join("], [")}]`));
    } else if (m[5] !== undefined) {
      // Resolve to the number the float actually carries; emitting the bare
      // label suffix would render "Fig. dynamics" in running text.
      out.push(new TextRun(LABELS[m[5]] || m[5].replace(/^(fig|tab|sec):/, "")));
    }
    last = re.lastIndex;
  }
  if (last < text.length) out.push(new TextRun(plain(text.slice(last))));
  return out.length ? out : [new TextRun("")];
}

/* ----------------------------------------------------------- tex utilities */
function braceArg(s, from) {
  // Return the balanced {...} argument starting at or after `from`.
  const start = s.indexOf("{", from);
  if (start === -1) return null;
  let depth = 0;
  for (let i = start; i < s.length; i++) {
    if (s[i] === "{") depth++;
    else if (s[i] === "}" && --depth === 0)
      return { body: s.slice(start + 1, i), end: i + 1 };
  }
  return null;
}

function environments(tex, name) {
  const out = [];
  const re = new RegExp(`\\\\begin\\{${name}\\}([\\s\\S]*?)\\\\end\\{${name}\\}`, "g");
  let m;
  while ((m = re.exec(tex)) !== null) out.push({ body: m[1], index: m.index });
  return out;
}

/* --------------------------------------------------------------- the build */
function buildTable(body, refs) {
  // Parse a tabular: strip rules, split on \\ then &.
  const inner = body.replace(/\\(top|mid|bottom)rule/g, "")
    .replace(/\\begin\{tabular\}\{[^}]*\}/, "")
    .replace(/\\end\{tabular\}/, "");
  const rows = inner.split(/\\\\/).map((r) => r.trim()).filter(Boolean);
  if (!rows.length) return null;
  const cells = rows.map((r) =>
    r.split("&").map((c) => c.replace(/\\multicolumn\{\d+\}\{[^}]*\}/, "").trim()));
  const width = Math.max(...cells.map((c) => c.length));
  const border = { style: BorderStyle.SINGLE, size: 2, color: "000000" };
  return new Table({
    width: { size: 100, type: WidthType.PERCENTAGE },
    rows: cells.map((row, ri) => new TableRow({
      children: Array.from({ length: width }, (_, ci) => new TableCell({
        borders: ri === 0
          ? { top: border, bottom: border, left: { style: BorderStyle.NONE }, right: { style: BorderStyle.NONE } }
          : { top: { style: BorderStyle.NONE }, bottom: ri === cells.length - 1 ? border : { style: BorderStyle.NONE }, left: { style: BorderStyle.NONE }, right: { style: BorderStyle.NONE } },
        children: [new Paragraph({
          style: ri === 0 ? "tablehead" : "tablecopy",
          alignment: ci === 0 ? AlignmentType.LEFT : AlignmentType.RIGHT,
          children: runs(row[ci] || "", refs),
        })],
      })),
    })),
  });
}

function convert(tex, refs, singleColumn) {
  const front = [];   // title, authors, abstract, keywords
  const body = [];

  // ---- front matter -------------------------------------------------------
  const title = braceArg(tex, tex.indexOf("\\title"));
  front.push(new Paragraph({
    style: "papertitle",
    children: runs((title ? title.body : "").replace(/\\\\/g, " "), refs),
  }));
  const authorBlock = braceArg(tex, tex.indexOf("\\author"));
  if (authorBlock) {
    const name = /\\IEEEauthorblockN\{([^}]*)\}/.exec(authorBlock.body);
    const aff = /\\IEEEauthorblockA\{([\s\S]*?)\}\s*\}?\s*$/.exec(authorBlock.body);
    if (name) front.push(new Paragraph({ style: "Author", children: runs(name[1], refs) }));
    if (aff) {
      aff[1].split(/\\\\/).map((l) => l.trim()).filter(Boolean).forEach((l) =>
        front.push(new Paragraph({ style: "Affiliation", children: runs(l, refs) })));
    }
  }
  const abs = environments(tex, "abstract")[0];
  if (abs) {
    front.push(new Paragraph({
      style: "Abstract",
      children: [new TextRun({ text: "Abstract—", bold: true, italics: true }),
        ...runs(abs.body.trim(), refs)],
    }));
  }
  const kw = environments(tex, "IEEEkeywords")[0];
  if (kw) {
    front.push(new Paragraph({
      style: "Keywords",
      children: [new TextRun({ text: "Index Terms—", bold: true, italics: true }),
        ...runs(kw.body.trim(), refs)],
    }));
  }

  // ---- body ---------------------------------------------------------------
  const start = tex.indexOf("\\section");
  const end = tex.indexOf("\\begin{thebibliography}");
  let src = tex.slice(start, end === -1 ? undefined : end);

  // Pull float environments out first, replacing them with placeholders so
  // that paragraph splitting does not tear them apart.
  const floats = [];
  src = src.replace(/\\begin\{(figure|table)\}[\s\S]*?\\end\{\1\}/g, (blk, kind) => {
    floats.push({ kind, blk });
    return `\n\n FLOAT${floats.length - 1} \n\n`;
  });

  for (const chunk of src.split(/\n\s*\n/)) {
    const t = chunk.trim();
    if (!t) continue;

    const ph = /^ FLOAT(\d+) $/.exec(t);
    if (ph) {
      const { kind, blk } = floats[+ph[1]];
      const cap = braceArg(blk, blk.indexOf("\\caption"));
      if (kind === "figure") {
        const img = /\\includegraphics\[[^\]]*\]\{([^}]+)\}/.exec(blk);
        if (img) {
          const file = path.join(FIGS, img[1] + ".png");
          if (fs.existsSync(file)) {
            // Column width at A4 with IEEE margins is ~3.4in; PNGs are 200dpi.
            body.push(new Paragraph({
              alignment: AlignmentType.CENTER,
              children: [new ImageRun({
                data: fs.readFileSync(file), type: "png",
                transformation: { width: 320, height: Math.round(320 * imgRatio(file)) },
              })],
            }));
          }
        }
        const n = body.filter((p) => p.__fig).length + 1;
        const p = new Paragraph({
          style: "figurecaption",
          children: [new TextRun({ text: `Fig. ${figCount(body)}. ` }),
            ...runs(cap ? cap.body : "", refs)],
        });
        p.__fig = true;
        body.push(p);
      } else {
        const p = new Paragraph({
          style: "tablehead",
          alignment: AlignmentType.CENTER,
          children: [new TextRun({ text: `TABLE ${roman(tabCount(body))}. ` }),
            ...runs(cap ? cap.body : "", refs)],
        });
        p.__tab = true;
        body.push(p);
        const tbl = buildTable(blk, refs);
        if (tbl) body.push(tbl);
      }
      continue;
    }

    const sec = /^\\section\*?\{/.test(t) ? braceArg(t, 0) : null;
    if (sec) {
      body.push(new Paragraph({ style: "Heading1", children: runs(sec.body, refs) }));
      const rest = t.slice(sec.end).trim();
      if (rest) body.push(new Paragraph({ style: "BodyText", children: runs(rest, refs) }));
      continue;
    }
    const sub = /^\\subsection\*?\{/.test(t) ? braceArg(t, 0) : null;
    if (sub) {
      body.push(new Paragraph({ style: "Heading2", children: runs(sub.body, refs) }));
      const rest = t.slice(sub.end).trim();
      if (rest) body.push(new Paragraph({ style: "BodyText", children: runs(rest, refs) }));
      continue;
    }
    if (/^\\begin\{itemize\}/.test(t)) {
      t.replace(/\\begin\{itemize\}|\\end\{itemize\}/g, "")
        .split(/\\item/).map((i) => i.trim()).filter(Boolean)
        .forEach((i) => body.push(new Paragraph({
          style: "BodyText", bullet: { level: 0 }, children: runs(i, refs),
        })));
      continue;
    }
    if (/^\\(label|graphicspath|newcommand|maketitle)/.test(t)) continue;
    body.push(new Paragraph({ style: "BodyText", children: runs(t, refs) }));
  }

  // ---- references ---------------------------------------------------------
  body.push(new Paragraph({ style: "Heading1", children: [new TextRun("References")] }));
  const bib = environments(tex, "thebibliography")[0];
  if (bib) {
    const items = bib.body.split(/\\bibitem\{[^}]*\}/).slice(1);
    items.forEach((it, i) => body.push(new Paragraph({
      style: "references",
      children: [new TextRun(`[${i + 1}] `), ...runs(it.trim(), refs)],
    })));
  }
  return { front, body };
}

const figCount = (b) => b.filter((p) => p.__fig).length + 1;
const tabCount = (b) => b.filter((p) => p.__tab).length + 1;
const roman = (n) => ["", "I", "II", "III", "IV", "V", "VI", "VII", "VIII"][n] || String(n);

/** Aspect ratio from the PNG header, so figures are not distorted. */
function imgRatio(file) {
  const buf = fs.readFileSync(file);
  const w = buf.readUInt32BE(16), h = buf.readUInt32BE(20);
  return h / w;
}

/* --------------------------------------------------------------------- main */
(async function main() {
  const singleColumn = process.argv.includes("--single-column");
  if (!fs.existsSync(TEX)) throw new Error(`missing ${TEX}`);
  if (!fs.existsSync(path.join(FIGS, "fig1_dynamics.png")))
    throw new Error("PNG figures missing - run paper/make_figures.py first");

  const tex = fs.readFileSync(TEX, "utf8");
  const refs = [...tex.matchAll(/\\bibitem\{([^}]*)\}/g)].map((m) => m[1]);

  // Inherit IEEE's paragraph styles so `papertitle`, `Abstract`, `Heading1`
  // and the rest resolve to the template's real formatting.
  let externalStyles;
  if (fs.existsSync(TEMPLATE)) {
    const zip = await JSZip.loadAsync(fs.readFileSync(TEMPLATE));
    externalStyles = await zip.file("word/styles.xml").async("string");
  } else {
    console.warn("warning: conference-template-a4.docx not found; output will "
      + "not carry IEEE formatting");
  }

  // Number the floats in source order, so \ref resolves the way LaTeX would.
  let nFig = 0, nTab = 0;
  for (const m of tex.matchAll(/\\begin\{(figure|table)\}[\s\S]*?\\end\{\1\}/g)) {
    const lbl = /\\label\{([^}]*)\}/.exec(m[0]);
    if (lbl) LABELS[lbl[1]] = m[1] === "figure" ? String(++nFig) : roman(++nTab);
  }
  // Section labels resolve to their section number.
  let nSec = 0;
  for (const m of tex.matchAll(/\\section\*?\{[^}]*\}\s*\n?\\label\{([^}]*)\}/g)) {
    LABELS[m[1]] = roman(++nSec);
  }

  const { front, body } = convert(tex, refs, singleColumn);

  // A4, IEEE margins. Front matter spans the page; the body is two columns.
  const page = { size: { width: 11906, height: 16838 },
    margin: { top: 720, right: 900, bottom: 720, left: 900 } };
  const sections = singleColumn
    ? [{ properties: { page }, children: [...front, ...body] }]
    : [
        { properties: { page, column: { count: 1 } }, children: front },
        { properties: { type: SectionType.CONTINUOUS, page,
            column: { count: 2, space: 360, equalWidth: true } }, children: body },
      ];

  const doc = new Document({ externalStyles, sections });
  fs.writeFileSync(OUT, await Packer.toBuffer(doc));

  const kb = (fs.statSync(OUT).size / 1024).toFixed(0);
  console.log(`wrote ${OUT}  (${kb} KB, ${singleColumn ? 1 : 2}-column, `
    + `template=${externalStyles ? "conference-template-a4.docx" : "none"})`);
  console.log(`  front matter: ${front.length} paragraphs`);
  console.log(`  body: ${body.length} blocks, ${refs.length} references`);
  console.log("  note: maths is Unicode text, not native Word equations - use "
    + "make_docx.py if the equations must be editable in Word.");
})().catch((e) => { console.error(e); process.exit(1); });

File: paper/make_docx.py
"""Convert the IEEE LaTeX paper to .docx using the official IEEE template.

Pandoc handles structure, tables, maths and cross-references. Everything below
covers what pandoc cannot know about.

**The template supplies the look.** ``conference-template-a4.docx`` is passed as
pandoc's reference document, so A4 page size, margins, fonts and the IEEE
paragraph styles come from IEEE rather than from us.

**Style names have to be remapped.** Pandoc emits its own style identifiers
(``Title``, ``ImageCaption``, ...) which are not the ones the IEEE template
defines (``papertitle``, ``figurecaption``, ...). Without the remapping the
document silently falls back to Normal and looks nothing like the template,
which is easy to miss because no error is raised.

**Sectioning has to be rebuilt.** Pandoc copies a single section layout from
the reference document and picks the wrong one --- the template's final,
single-column section. Real IEEE layout is single-column title and abstract
followed by a two-column body, which is two sections separated by a continuous
break. That break is inserted here.

**Figures and macros.** The LaTeX build uses vector PDFs, which Word will not
render, so graphics paths are rewritten to the PNG twins that
``make_figures.py`` also emits. The custom ``\\evrate`` and ``\\osc`` macros are
expanded, since pandoc drops unknown macros silently and would leave gaps
exactly where the two headline metrics are named.

The ``.tex`` remains the source of record: regenerate rather than editing the
``.docx``, or the two will drift apart.

Usage
-----
    python paper/make_docx.py
    python paper/make_docx.py --single-column   # easier to comment on
"""

from __future__ import annotations

import argparse
import os
import re
import tempfile
import zipfile
from pathlib import Path

HERE = Path(__file__).resolve().parent
TEX = HERE / "main.tex"
OUT = HERE / "main.docx"
TEMPLATE = HERE / "conference-template-a4.docx"

W = "http://schemas.openxmlformats.org/wordprocessingml/2006/main"

# Preamble macros pandoc will not expand on its own.
MACROS = {r"\evrate": "*E*~tail~", r"\osc": "\u03a9"}

# Pandoc's style identifiers -> the identifiers the IEEE template defines.
# Anything left unmapped falls back to Normal and loses the template's look.
STYLE_MAP = {
    "Title": "papertitle",
    "Author": "Author",
    "ImageCaption": "figurecaption",
    "TableCaption": "tablehead",
    "Bibliography": "references",
    "FirstParagraph": "BodyText",
    "Compact": "BodyText",
    "SourceCode": "BodyText",
}


def preprocess(tex: str) -> str:
    """Make the LaTeX pandoc-friendly without changing its meaning."""
    # Word cannot display the vector PDFs; the PNG twins are already built.
    tex = re.sub(r"(\\includegraphics\[[^\]]*\]\{)([^}]+)(\})",
                 lambda m: m.group(1) + m.group(2) + ".png" + m.group(3), tex)
    # Strip the \newcommand definitions before expanding uses of those macros:
    # expanding first would rewrite the definitions into invalid LaTeX.
    for name in MACROS:
        tex = re.sub(r"\\newcommand\{" + re.escape(name) + r"\}\{[^\n]*\}\n",
                     "", tex)
    # Longest first, so a macro that prefixes another cannot consume it.
    for name in sorted(MACROS, key=len, reverse=True):
        tex = tex.replace(name + "{}", MACROS[name]).replace(name, MACROS[name])
    # Packages that only affect PDF typesetting.
    for pkg in ("balance", "hyperref"):
        tex = re.sub(r"\\usepackage(\[[^\]]*\])?\{" + pkg + r"\}\n", "", tex)
    return tex.replace("\\balance\n", "")


def _sect(columns: int, continuous: bool) -> str:
    """A section-properties block on the template's A4 geometry."""
    typ = '<w:type w:val="continuous"/>' if continuous else ""
    cols = (f'<w:cols w:num="{columns}" w:space="18pt" w:equalWidth="1"/>'
            if columns > 1 else '<w:cols w:space="36pt"/>')
    return (f'<w:sectPr xmlns:w="{W}">{typ}'
            f'<w:pgSz w:w="595.30pt" w:h="841.90pt" w:code="9"/>'
            f'<w:pgMar w:top="36pt" w:right="45pt" w:bottom="36pt" '
            f'w:left="45pt" w:header="0pt" w:footer="0pt" w:gutter="0pt"/>'
            f'{cols}</w:sectPr>')


def _apply_styles(xml: str) -> tuple[str, dict]:
    counts: dict[str, int] = {}

    def sub(m: re.Match) -> str:
        old = m.group(1)
        new = STYLE_MAP.get(old)
        if new is None:
            return m.group(0)
        counts[f"{old} -> {new}"] = counts.get(f"{old} -> {new}", 0) + 1
        return m.group(0).replace(f'w:val="{old}"', f'w:val="{new}"')

    return re.sub(r'<w:pStyle w:val="([^"]+)"\s*/?>', sub, xml), counts


def _split_sections(xml: str, columns: int) -> tuple[str, bool]:
    """Single-column front matter, then a two-column body.

    A ``sectPr`` inside a paragraph governs the section *ending* at that
    paragraph, so placing a single-column block in the last front-matter
    paragraph leaves the two-column block at the end of the body to govern
    everything after it.
    """
    if columns < 2:
        return xml, False
    # The keywords paragraph is the last of the front matter.
    paras = list(re.finditer(r"<w:p\b.*?</w:p>", xml, re.S))
    anchor = None
    for p in paras:
        if 'w:val="Keywords"' in p.group(0) or "Index Terms" in p.group(0):
            anchor = p
    if anchor is None:                       # fall back to the abstract
        for p in paras:
            if 'w:val="Abstract"' in p.group(0):
                anchor = p
    if anchor is None:
        return xml, False

    block = anchor.group(0)
    front = _sect(1, continuous=False)
    if "<w:pPr>" in block:
        patched = block.replace("<w:pPr>", "<w:pPr>" + front, 1)
    else:
        patched = block.replace("<w:p ", "<w:p ", 1)
        patched = re.sub(r"(<w:p\b[^>]*>)", r"\1<w:pPr>" + front + "</w:pPr>",
                         patched, count=1)
    return xml[:anchor.start()] + patched + xml[anchor.end():], True


def _set_body_section(xml: str, sect: str) -> str:
    """Replace only the body-level ``sectPr``, the one just before ``</w:body>``.

    Done by index rather than by regex: a lazy pattern anchored on
    ``</w:body>`` will happily start at the front-matter section inserted
    above and swallow the entire document between the two, which produces a
    valid but empty file rather than an error.
    """
    end = xml.rfind("</w:body>")
    if end == -1:
        return xml
    start = xml.rfind("<w:sectPr", 0, end)
    # Only a *body-level* sectPr sits directly before </w:body>; a paragraph
    # one would be followed by </w:pPr></w:p>.
    if start != -1 and "</w:p>" not in xml[start:end]:
        return xml[:start] + sect + xml[end:]
    return xml[:end] + sect + xml[end:]


def postprocess(docx: Path, columns: int) -> dict:
    tmp = docx.with_suffix(".tmp.docx")
    report: dict = {}
    with zipfile.ZipFile(docx) as zin, zipfile.ZipFile(
            tmp, "w", zipfile.ZIP_DEFLATED) as zout:
        for item in zin.infolist():
            data = zin.read(item.filename)
            if item.filename == "word/document.xml":
                xml = data.decode("utf-8")
                xml, report["styles"] = _apply_styles(xml)
                xml, report["split"] = _split_sections(xml, columns)
                xml = _set_body_section(
                    xml, _sect(columns, continuous=bool(report["split"])))
                data = xml.encode("utf-8")
            zout.writestr(item, data)
    tmp.replace(docx)
    return report


def build(columns: int) -> tuple[Path, dict]:
    import pypandoc

    tex = preprocess(TEX.read_text(encoding="utf-8"))
    args = [
        # os.pathsep, not ":" -- on Windows a colon splits the drive letter and
        # every image silently fails to resolve, with no error raised.
        "--resource-path=" + os.pathsep.join([str(HERE), str(HERE / "figures")]),
        "--from=latex+raw_tex",
        "--number-sections",
        "--wrap=preserve",
    ]
    if TEMPLATE.exists():
        args.append(f"--reference-doc={TEMPLATE}")
    with tempfile.TemporaryDirectory() as td:
        staged = Path(td) / "main.tex"
        staged.write_text(tex, encoding="utf-8")
        pypandoc.convert_file(str(staged), "docx", format="latex",
                              outputfile=str(OUT), extra_args=args)
    return OUT, postprocess(OUT, columns)


def main() -> None:
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--single-column", action="store_true")
    args = ap.parse_args()

    if not (HERE / "figures" / "fig1_dynamics.png").exists():
        raise SystemExit("PNG figures missing - run paper/make_figures.py first")
    if not TEMPLATE.exists():
        print(f"warning: {TEMPLATE.name} not found; falling back to pandoc "
              "defaults, which will not look like an IEEE submission")

    columns = 1 if args.single_column else 2
    out, report = build(columns)
    print(f"wrote {out}  ({out.stat().st_size/1024:.0f} KB, "
          f"{columns}-column, template={TEMPLATE.name if TEMPLATE.exists() else 'none'})")
    for k, v in sorted(report.get("styles", {}).items()):
        print(f"  restyled {v:3d}x  {k}")
    print(f"  single-column front matter: {report.get('split')}")
    print("The .tex stays the source of record: regenerate rather than "
          "editing the .docx.")


if __name__ == "__main__":
    main()
